File: src/widgets/fields/inputLogicFields.ts
import type { ModuleType,
  BranchGateData,
  DatePickerData,
  FormulaData,
  NumberInputData,
  StatusData,
  TextInputData,
  ToggleData,
} from '../../types/spatial'
import type { FieldDescriptor } from '../contracts/fields'
import { num, text, bool, formulaValue, daysUntil } from './valueHelpers'

/** Input and logic widget fields (text_input … date_picker). Extracted verbatim from fields.ts; field order IS port-slot order — never reorder within an entry. */
export const INPUT_LOGIC_FIELDS = {
  text_input: [
    {
      key: 'value',
      label: 'Text value',
      valueType: 'text',
      get: (d) => (d as TextInputData).value,
      set: (d, v) => ({ ...(d as TextInputData), value: text(v) }),
    },
    {
      key: 'has_value',
      label: 'Has value',
      valueType: 'boolean',
      get: (d) => (d as TextInputData).value.trim().length > 0,
    },
  ],
  number_input: [
    {
      key: 'value',
      label: 'Number value',
      valueType: 'number',
      get: (d) => (d as NumberInputData).value,
      set: (d, v) => {
        const nd = d as NumberInputData
        const min = Math.min(nd.min, nd.max)
        const max = Math.max(nd.min, nd.max)
        return { ...nd, value: Math.min(max, Math.max(min, num(v))) }
      },
    },
  ],
  toggle: [
    {
      key: 'value',
      label: 'On / off',
      valueType: 'boolean',
      get: (d) => (d as ToggleData).value,
      set: (d, v) => ({ ...(d as ToggleData), value: bool(v) }),
    },
  ],
  branch_gate: [
    {
      key: 'value',
      label: 'True branch',
      valueType: 'boolean',
      get: (d) => (d as BranchGateData).value,
      set: (d, v) => ({ ...(d as BranchGateData), value: bool(v) }),
    },
    {
      key: 'inverse',
      label: 'False branch',
      valueType: 'boolean',
      get: (d) => !(d as BranchGateData).value,
    },
  ],
  formula: [
    {
      key: 'a',
      label: 'Input A',
      valueType: 'number',
      get: (d) => (d as FormulaData).a,
      set: (d, v) => ({ ...(d as FormulaData), a: num(v) }),
    },
    {
      key: 'b',
      label: 'Input B',
      valueType: 'number',
      get: (d) => (d as FormulaData).b,
      set: (d, v) => ({ ...(d as FormulaData), b: num(v) }),
    },
    {
      key: 'result',
      label: 'Result',
      valueType: 'number',
      get: (d) => formulaValue(d as FormulaData),
    },
  ],
  status: [
    {
      key: 'status',
      label: 'Status',
      valueType: 'text',
      get: (d) => (d as StatusData).value,
      set: (d, v) => {
        const value = text(v)
        const legal = ['not_started', 'in_progress', 'blocked', 'done'] as const
        return legal.includes(value as (typeof legal)[number])
          ? { ...(d as StatusData), value: value as StatusData['value'] }
          : d
      },
    },
    {
      key: 'progress',
      label: 'Progress %',
      valueType: 'number',
      unit: 'percent',
      get: (d) => {
        const value = (d as StatusData).value
        return value === 'done' ? 100 : value === 'in_progress' || value === 'blocked' ? 50 : 0
      },
    },
    {
      key: 'complete',
      label: 'Complete',
      valueType: 'boolean',
      get: (d) => (d as StatusData).value === 'done',
    },
  ],
  date_picker: [
    {
      key: 'date',
      label: 'Date',
      valueType: 'text',
      unit: 'date_iso',
      get: (d) => (d as DatePickerData).date,
      set: (d, v) => ({ ...(d as DatePickerData), date: text(v) }),
    },
    {
      key: 'days_until',
      label: 'Days until',
      valueType: 'number',
      unit: 'count',
      get: (d) => daysUntil((d as DatePickerData).date),
      timeSensitive: true,
    },
    {
      key: 'is_due',
      label: 'Is due',
      valueType: 'boolean',
      get: (d) => Boolean((d as DatePickerData).date) && daysUntil((d as DatePickerData).date) <= 0,
      timeSensitive: true,
    },
  ],
} satisfies Partial<Record<ModuleType, FieldDescriptor[]>>

File: apple/Sources/GrovepadCore/Fields/InputLogicFields.swift
import Foundation

// ---------------------------------------------------------------------------
// Input and logic widget fields (`widgets/fields/inputLogicFields.ts`):
// text_input, number_input, toggle, status. `formula` and `date_picker`
// depend on skin models and are deferred (see AGENTS.md). Field order IS
// port-slot order.
// ---------------------------------------------------------------------------

enum InputLogicFields {
    static let legalStatuses: Set<String> = ["not_started", "in_progress", "blocked", "done"]

    static let tables: [String: [FieldDescriptor]] = [
        "text_input": [
            FieldDescriptor(
                key: "value", label: "Text value", valueType: .text,
                get: { $0.fieldValue("value", default: .text("")) },
                set: { data, value, _ in data.assigning("value", .string(text(value))) }
            ),
            FieldDescriptor(
                key: "has_value", label: "Has value", valueType: .boolean,
                get: { data in .bool(!data.trimmed("value").isEmpty) }
            ),
        ],
        "number_input": [
            FieldDescriptor(
                key: "value", label: "Number value", valueType: .number,
                get: { $0.fieldValue("value", default: .number(0)) },
                set: { data, value, _ in
                    let lo = Swift.min(data.dbl("min"), data.dbl("max"))
                    let hi = Swift.max(data.dbl("min"), data.dbl("max"))
                    return data.assigning("value", .number(Swift.min(hi, Swift.max(lo, num(value)))))
                }
            ),
        ],
        "toggle": [
            FieldDescriptor(
                key: "value", label: "On / off", valueType: .boolean,
                get: { $0.fieldValue("value", default: .bool(false)) },
                set: { data, value, _ in data.assigning("value", .bool(bool(value))) }
            ),
        ],
        "status": [
            FieldDescriptor(
                key: "status", label: "Status", valueType: .text,
                get: { $0.fieldValue("value", default: .text("")) },
                set: { data, value, _ in
                    let status = text(value)
                    return legalStatuses.contains(status) ? data.assigning("value", .string(status)) : data
                }
            ),
            FieldDescriptor(
                key: "progress", label: "Progress %", valueType: .number, unit: .percent,
                get: { data in
                    let status = data.string("value")
                    return .number(status == "done" ? 100 : (status == "in_progress" || status == "blocked") ? 50 : 0)
                }
            ),
            FieldDescriptor(
                key: "complete", label: "Complete", valueType: .boolean,
                get: { data in .bool(data.string("value") == "done") }
            ),
        ],
    ]
}
